File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/security/PrivilegedCommandBeanPostProcessor.java
package com.github.bumblebee.security;

import com.github.bumblebee.security.exception.UnprivilegedExecutionException;
import com.github.bumblebee.security.roles.service.UserRolesService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.aop.MethodBeforeAdvice;
import org.springframework.aop.framework.ProxyFactory;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.config.BeanPostProcessor;
import org.springframework.stereotype.Component;
import telegram.domain.Update;
import telegram.polling.UpdateHandler;

import java.lang.reflect.Method;
import java.text.MessageFormat;
import java.util.HashMap;
import java.util.Map;

@Component
public class PrivilegedCommandBeanPostProcessor implements BeanPostProcessor {

    private static final Logger log = LoggerFactory.getLogger(PrivilegedCommandBeanPostProcessor.class);

    private final Map<String, Class> beans = new HashMap<>();
    private final Method updateHandlerMethod;

    private final UserRolesService rolesService;

    @Autowired
    public PrivilegedCommandBeanPostProcessor(UserRolesService rolesService) {
        this.rolesService = rolesService;
        updateHandlerMethod = UpdateHandler.class.getMethods()[0];
    }

    @Override
    public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
        Class<?> clazz = bean.getClass();
        if (clazz.isAnnotationPresent(PrivilegedCommand.class)) {
            if (!UpdateHandler.class.isAssignableFrom(clazz)) {
                throw new IllegalStateException(clazz.getName() + "is not an UpdateHandler");
            }
            beans.put(beanName, clazz);
        }
        return bean;
    }

    @Override
    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
        Class clazz = beans.get(beanName);
        if (clazz == null) {
            return bean;
        }

        log.info("Instantiating proxy for {}", clazz);

        ProxyFactory proxyFactory = new ProxyFactory(bean);
        proxyFactory.addAdvice((MethodBeforeAdvice) (method, args, target) -> {

            if (updateHandlerMethod.equals(method)) {
                Update update = (Update) args[0];
                UserRole role = ((PrivilegedCommand) clazz.getAnnotation(PrivilegedCommand.class)).role();

                ensurePrivilege(target, update, role);
            }
        });
        return proxyFactory.getProxy();
    }

    private void ensurePrivilege(Object bean, Update update, UserRole role) {
        Long userId = update.getMessage().getFrom().getId();
        log.info("Executing privileged from: {}", userId);

        if (!rolesService.hasPrivilege(userId, role)) {

            if (bean instanceof UnauthorizedRequestAware) {
                ((UnauthorizedRequestAware) bean).onUnauthorizedRequest(update);
            }

            throw new UnprivilegedExecutionException(
                    MessageFormat.format("Role '{0}' required. User id: {1}", role, userId));
        }
    }
}

File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/security/PrivilegedCommandBeanPostProcessor.kt
package com.github.bumblebee.security

import com.github.bumblebee.bot.consumer.UpdateHandler
import com.github.bumblebee.security.exception.UnprivilegedExecutionException
import com.github.bumblebee.security.roles.service.UserRolesService
import com.github.bumblebee.util.logger
import com.github.telegram.domain.Update
import org.springframework.aop.MethodBeforeAdvice
import org.springframework.aop.framework.ProxyFactory
import org.springframework.beans.BeansException
import org.springframework.beans.factory.config.BeanPostProcessor
import org.springframework.stereotype.Component
import java.lang.reflect.Method
import java.util.*

@Component
class PrivilegedCommandBeanPostProcessor(private val rolesService: UserRolesService) : BeanPostProcessor {

    private val beans = HashMap<String, Class<*>>()
    private val updateHandlerMethod: Method = UpdateHandler::class.java.methods.first()

    @Throws(BeansException::class)
    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any {
        val clazz = bean.javaClass
        if (clazz.isAnnotationPresent(PrivilegedCommand::class.java)) {
            if (!UpdateHandler::class.java.isAssignableFrom(clazz)) {
                throw IllegalStateException(clazz.name + "is not an UpdateHandler")
            }
            beans[beanName] = clazz
        }
        return bean
    }

    @Throws(BeansException::class)
    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any {
        val clazz = beans[beanName] ?: return bean

        log.info("Instantiating proxy for {}", clazz)

        return with(ProxyFactory(bean)) {
            addAdvice(MethodBeforeAdvice { method, args, target ->
                if (updateHandlerMethod == method) {
                    val update = args[0] as Update
                    val role = (clazz.getAnnotation(PrivilegedCommand::class.java) as PrivilegedCommand).role

                    ensurePrivilege(target, update, role)
                }
            })
            proxy
        }
    }

    private fun ensurePrivilege(bean: Any?, update: Update, role: UserRole) {
        val userId = update.message?.from?.id
        log.info("Executing privileged from: {}", userId)

        if (userId == null || !rolesService.hasPrivilege(userId, role)) {
            (bean as? UnauthorizedRequestAware)?.onUnauthorizedRequest(update)
            throw UnprivilegedExecutionException("Role '$role' required. User id: $userId")
        }
    }

    companion object {
        private val log = logger<PrivilegedCommandBeanPostProcessor>()
    }
}
